README.md
# Capacitor Native Biometric

Use biometrics confirm device owner presence or authenticate users. A couple of methods are provided to handle user credentials. These are securely stored using Keychain (iOS) and Keystore (Android).

## Installation (Only supports Capacitor 3 and 4)

- `npm i capacitor-native-biometric`

## Usage

```ts
import { NativeBiometric } from "capacitor-native-biometric";

async performBiometricVerificatin(){
  const result = await NativeBiometric.isAvailable();

  if(!result.isAvailable) return;

  const isFaceID = result.biometryType == BiometryType.FACE_ID;

  const verified = await NativeBiometric.verifyIdentity({
    reason: "For easy log in",
    title: "Log in",
    subtitle: "Maybe add subtitle here?",
    description: "Maybe a description too?",
  })
    .then(() => true)
    .catch(() => false);

  if(!verified) return;

  const credentials = await NativeBiometric.getCredentials({
    server: "www.example.com",
  });
}

// Save user's credentials
NativeBiometric.setCredentials({
  username: "username",
  password: "password",
  server: "www.example.com",
}).then();

// Delete user's credentials
NativeBiometric.deleteCredentials({
  server: "www.example.com",
}).then();
```

## Methods

| Method                                                | Default | Type                       | Description                                                                                   |
| ----------------------------------------------------- | ------- | -------------------------- | --------------------------------------------------------------------------------------------- |
| `isAvailable(options?: IsAvailableOptions)`           |         | `Promise<AvailableResult>` | Gets available biometrics                                                                     |
| `verifyIdentity(options?: BiometricOptions)`          |         | `Promise<void>`            | Shows biometric prompt                                                                        |
| `setCredentials(options: SetCredentialOptions)`       |         | `Promise<void>`            | Securely stores user's credentials in Keychain (iOS) or encypts them using Keystore (Android) |
| `getCredentials(options: GetCredentialOptions)`       |         | `Promise<Credentials>`     | Retrieves user's credentials if any                                                           |
| `deleteCredentials(options: DeleteCredentialOptions)` |         | `Promise<void>`            | Removes user's credentials if any                                                             |

## Interfaces

### IsAvailableOptions

| Properties     | Default | Type      | Description                                                               |
| -------------- | ------- | --------- | ------------------------------------------------------------------------- |
| `useFallback?` |         | `boolean` | Specifies if the device should fallback to using passcode authentication. |

### AvailableResult

| Properties     | Default | Type           | Description                                              |
| -------------- | ------- | -------------- | -------------------------------------------------------- |
| `isAvailable`  |         | `boolean`      | Specifies if the devices has biometric enrollment        |
| `biometryType` |         | `BiometryType` | Specifies the available biometric hardware on the device |
| `errorCode?`   |         | `number`       | Biometric Auth Error Code                                |

### BiometryType - enum

| Properties            | Description                                                                                                                   |
| --------------------- | ----------------------------------------------------------------------------------------------------------------------------- |
| `NONE`                | There is no biometry available                                                                                                |
| `TOUCH_ID`            | TouchID is available (iOS)                                                                                                    |
| `FACE_ID`             | FaceID is available (iOS)                                                                                                     |
| `FINGERPRINT`         | Fingerprint is available (Android)                                                                                            |
| `FACE_AUTHENTICATION` | Face Authentication is available (Android)                                                                                    |
| `IRIS_AUTHENTICATION` | Iris Authentication is available (Android)                                                                                    |
| `MULTIPLE`            | Returned when device has multiple biometric features. Currently there is no way of knowing which one is being used. (Android) |

### BiometricOptions

| Properties            | Default                        | Type      | Description                                                                                               |
| --------------------- | ------------------------------ | --------- | --------------------------------------------------------------------------------------------------------- |
| `reason?`             | "For biometric authentication" | `string`  | Reason for requesting authentication in iOS. Displays in the authentication dialog presented to the user. |
| `title?`              | "Authenticate"                 | `string`  | Title for the Android prompt                                                                              |
| `subtitle?`           |                                | `string`  | Subtitle for the Android prompt                                                                           |
| `description?`        |                                | `string`  | Description for the Android prompt                                                                        |
| `negativeButtonText?` | "Cancel"                       | `string`  | Text for the negative button displayed on Android                                                         |
| `maxAttempts?`        | 1                              | `number`  | Limit the number of attempts a user can perform biometric authentication. (Android - Max 5)               |
| `useFallback?`        | `false`                        | `boolean` | Specifies if the device should fallback to using passcode authentication.                                 |

### Biometric Auth Errors

This is a plugin specific list of error codes that can be thrown on verifyIdentity failure, or set as a part of isAvailable. It consolidates Android and iOS specific Authentication Error codes into one combined error list.

| Code | Description                 | Platform                     |
| ---- | --------------------------- | ---------------------------- |
| 0    | Unknown Error               | Android, iOS                 |
| 1    | Biometrics Unavailable      | Android, iOS                 |
| 2    | User Lockout                | Android, iOS                 |
| 3    | Biometrics Not Enrolled     | Android, iOS                 |
| 4    | User Temporary Lockout      | Android (Lockout for 30sec)  |
| 10   | Authentication Failed       | Android, iOS                 |
| 11   | App Cancel                  | iOS                          |
| 12   | Invalid Context             | iOS                          |
| 13   | Not Interactive             | iOS                          |
| 14   | Passcode Not Set            | Android, iOS                 |
| 15   | System Cancel               | Android, iOS                 |
| 16   | User Cancel                 | Android, iOS                 |
| 17   | User Fallback               | Android, iOS                 |

### SetCredentialOptions

| Properties | Default | Type     | Description                                                                                                                                                             |
| ---------- | ------- | -------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `username` |         | `string` | The string used as the alias at the time of loggin in. It doesn't have to be a username. For example if you're using email to log in your users then provide the email. |
| `password` |         | `string` | The users' password                                                                                                                                                     |
| `server`   |         | `string` | Any string to identify the credentials object with                                                                                                                      |

### GetCredentialOptions

| Properties | Default | Type     | Description                                                                      |
| ---------- | ------- | -------- | -------------------------------------------------------------------------------- |
| `server`   |         | `string` | The string used to identify the credentials object when setting the credentials. |

### DeleteCredentialOptions

| Properties | Default | Type     | Description                                                                      |
| ---------- | ------- | -------- | -------------------------------------------------------------------------------- |
| `server`   |         | `string` | The string used to identify the credentials object when setting the credentials. |

### Credentials

| Properties | Default | Type     | Description                                                                 |
| ---------- | ------- | -------- | --------------------------------------------------------------------------- |
| `username` |         | `string` | The username returned from `getCredentials(options: GetCredentialOptions)`. |
| `password` |         | `string` | The password returned from `getCredentials(options: GetCredentialOptions)`. |

## Face ID (iOS)

To use FaceID Make sure to provide a value for NSFaceIDUsageDescription, otherwise your app may crash on iOS devices with FaceID.

This value is just the reason for using FaceID. You can add something like the following example to App/info.plist:

```xml
<key>NSFaceIDUsageDescription</key>
<string>For an easier and faster log in.</string>
```

## Biometric (Android)

To use android's BiometricPrompt api you must add the following permission to your AndroidManifest.xml:

```xml
<uses-permission android:name="android.permission.USE_BIOMETRIC">
```

And register the plugin by adding it to you MainActivity's onCreate (Not needed for Capacitor 3):

```java
import com.epicshaggy.biometric.NativeBiometric;

public class MainActivity extends BridgeActivity {
  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    // Initializes the Bridge
    this.init(savedInstanceState, new ArrayList<Class<? extends Plugin>>() {{
      // Additional plugins you've installed go here
      // Ex: add(TotallyAwesomePlugin.class);
      add(NativeBiometric.class);
    }});
  }
}
```

## Contributors

[Jonthia](https://github.com/jonthia)
[One Click Web Studio](https://github.com/oneclickwebstudio)

### Want to Contribute?

Learn about contributing [HERE](./CONTRIBUTING.md)

## Notes

Hasn't been tested on Android API level 22 or lower.


CHANGELOG.md
#### 4.1.3 (2022-08-19)

##### Bug Fixes

*  Remove AppTheme.NoAction bar fixes [#63](https://github.com/epicshaggy/capacitor-native-biometric/pull/63) ([1ec54f4f](https://github.com/epicshaggy/capacitor-native-biometric/commit/1ec54f4ff41111ae916a1e9900ddbb683c407a1a))

#### 4.1.2 (2022-08-11)

##### Bug Fixes

* **android:**  correctly determine biometry type ([65fe9c6f](https://github.com/epicshaggy/capacitor-native-biometric/commit/65fe9c6f92e315f5f893327d94e44a2425f50c20))

#### 4.1.1 (2022-08-10)

##### Bug Fixes

* **useFallback:**  Added to android ([393c5c4d](https://github.com/epicshaggy/capacitor-native-biometric/commit/393c5c4de8757c9849ae92d92350029e8144d4d5))

#### 4.1.0 (2022-08-10)

##### New Features

- **android:** Add option for retries on Android ([2907ef59](https://github.com/epicshaggy/capacitor-native-biometric/commit/2907ef59824ef0b2e3cb59c83708d713a2988708))

##### Bug Fixes

- **isAvailable:** biometryType always returned 0 ([6702badb](https://github.com/epicshaggy/capacitor-native-biometric/commit/6702badb3031906963949dd88e42cfe663b31009))
- AvailableResult type missing property ([3a4b4eec](https://github.com/epicshaggy/capacitor-native-biometric/commit/3a4b4eecc8220ca87f558fb12dd31eef81239088))
- **android:** typo on AuthActivity class ([37287659](https://github.com/epicshaggy/capacitor-native-biometric/commit/372876598ede6aa7bf113230dda7388854a0f473))

# 4.0.0

## What's Changed

- Capacitor 4 support by @kalimantos in https://github.com/epicshaggy/capacitor-native-biometric/pull/62

## New Contributors

- @kalimantos made their first contribution in https://github.com/epicshaggy/capacitor-native-biometric/pull/62

**Full Changelog**: https://github.com/epicshaggy/capacitor-native-biometric/compare/3.1.1...4.0.0

# 3.1.1

## What's Changed

- Added password fallback and error code response by @jonthia in https://github.com/epicshaggy/capacitor-native-biometric/pull/10
- Integration to Capacitor 3 by @oneclickwebstudio in https://github.com/epicshaggy/capacitor-native-biometric/pull/9
- Remove the WebPlugin Config Object by @Myles-Trevino in https://github.com/epicshaggy/capacitor-native-biometric/pull/18
- Set minimum iOS deployment target to 12.0 to match Capacitor 3 by @alexcroox in https://github.com/epicshaggy/capacitor-native-biometric/pull/13
- Update readme.md by @tolutronics in https://github.com/epicshaggy/capacitor-native-biometric/pull/22
- Remove label 'defaultValue' for Swift 5 compat by @squio in https://github.com/epicshaggy/capacitor-native-biometric/pull/52

## New Contributors

- @jonthia made their first contribution in https://github.com/epicshaggy/capacitor-native-biometric/pull/10
- @oneclickwebstudio made their first contribution in https://github.com/epicshaggy/capacitor-native-biometric/pull/9
- @Myles-Trevino made their first contribution in https://github.com/epicshaggy/capacitor-native-biometric/pull/18
- @alexcroox made their first contribution in https://github.com/epicshaggy/capacitor-native-biometric/pull/13
- @pzontrop made their first contribution in https://github.com/epicshaggy/capacitor-native-biometric/pull/21
- @tolutronics made their first contribution in https://github.com/epicshaggy/capacitor-native-biometric/pull/22
- @squio made their first contribution in https://github.com/epicshaggy/capacitor-native-biometric/pull/52

**Full Changelog**: https://github.com/epicshaggy/capacitor-native-biometric/commits/3.1.1


ios/Plugin/Plugin.swift
import Foundation
import Capacitor
import LocalAuthentication

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitor.ionicframework.com/docs/plugins/ios
 */

@objc(NativeBiometric)
public class NativeBiometric: CAPPlugin {
    
    struct Credentials {
        var username: String
        var password: String
    }
    
    enum KeychainError: Error{
        case noPassword
        case unexpectedPasswordData
        case duplicateItem
        case unhandledError(status: OSStatus)
    }
    
    typealias JSObject = [String:Any]
    
    @objc func isAvailable(_ call: CAPPluginCall) {
        let context = LAContext()
        var error: NSError?
        var obj = JSObject()
        
        obj["isAvailable"] = false
        obj["biometryType"] = 0

        let useFallback = call.getBool("useFallback", false)
        let policy = useFallback ? LAPolicy.deviceOwnerAuthentication : LAPolicy.deviceOwnerAuthenticationWithBiometrics
        
        if context.canEvaluatePolicy(policy, error: &error){
            switch context.biometryType {
                case .touchID:
                    obj["biometryType"] = 1
                case .faceID:
                    obj["biometryType"] = 2
                default:
                    obj["biomertryType"] = 0
            }
            
            obj["isAvailable"] = true
            call.resolve(obj)
        } else {
            guard let authError = error else {
                obj["errorCode"] = 0
                call.resolve(obj)
                return
            }
            var errorCode = 0
            switch authError.code {
                case LAError.biometryNotAvailable.rawValue:
                    errorCode = 1
                    
                case LAError.biometryLockout.rawValue:
                    errorCode = 2 //"Authentication could not continue because the user has been locked out of biometric authentication, due to failing authentication too many times."
                    
                case LAError.biometryNotEnrolled.rawValue:
                    errorCode = 3//message = "Authentication could not start because the user has not enrolled in biometric authentication."
                    
                default:
                    errorCode = 0 //"Did not find error code on LAError object"
            }
            obj["errorCode"] = errorCode
            call.resolve(obj)
        }
                        
    }
    
    @objc func verifyIdentity(_ call: CAPPluginCall){
        let context = LAContext()
        var canEvaluateError: NSError?

        let useFallback = call.getBool("useFallback", false)
        context.localizedFallbackTitle = useFallback ? nil : ""
        
        let policy = useFallback ? LAPolicy.deviceOwnerAuthentication : LAPolicy.deviceOwnerAuthenticationWithBiometrics
        
        if context.canEvaluatePolicy(policy, error: &canEvaluateError){
            
            let reason = call.getString("reason") ?? "For biometric authentication"

            context.evaluatePolicy(policy, localizedReason: reason) { (success, evaluateError) in
                
                if success {
                    call.resolve()
                }else{
                    var errorCode = "0"
                    guard let error = evaluateError
                    else {
                        call.reject("Biometrics Error", "0")
                        return
                    }
                    
                    switch error._code {
                    
                    case LAError.authenticationFailed.rawValue:
                        errorCode = "10"
                        
                    case LAError.appCancel.rawValue:
                        errorCode = "11"
                        
                    case LAError.invalidContext.rawValue:
                        errorCode = "12"
                        
                    case LAError.notInteractive.rawValue:
                        errorCode = "13"
                        
                    case LAError.passcodeNotSet.rawValue:
                        errorCode = "14"
                        
                    case LAError.systemCancel.rawValue:
                        errorCode = "15"
                        
                    case LAError.userCancel.rawValue:
                        errorCode = "16"
                        
                    case LAError.userFallback.rawValue:
                        errorCode = "17"

                    case LAError.biometryNotAvailable.rawValue:
                        errorCode = "1"
                
                    case LAError.biometryLockout.rawValue:
                        errorCode = "2" //"Authentication could not continue because the user has been locked out of biometric authentication, due to failing authentication too many times."
                
                    case LAError.biometryNotEnrolled.rawValue:
                        errorCode = "3" //message = "Authentication could not start because the user has not enrolled in biometric authentication."
                        
                    default:
                        errorCode = "0" // Biometrics unavailable
                    }                    
                    call.reject(error.localizedDescription, errorCode, error )
                }
                
            }
            
        }else{
            call.reject("Authentication not available")
        }
    }
    
    @objc func getCredentials(_ call: CAPPluginCall){
        guard let server = call.getString("server") else{
            call.reject("No server name was provided")
            return
        }
        do{
            let credentials = try getCredentialsFromKeychain(server)
            var obj = JSObject()
            obj["username"] = credentials.username
            obj["password"] = credentials.password
            call.resolve(obj)
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func setCredentials(_ call: CAPPluginCall){
        
        guard let server = call.getString("server"), let username = call.getString("username"), let password = call.getString("password") else {
            call.reject("Missing properties")
            return;
        }
        
        let credentials = Credentials(username: username, password: password)
        
        do{
            try storeCredentialsInKeychain(credentials, server)
            call.resolve()
        } catch KeychainError.duplicateItem {
            do {
                try updateCredentialsInKeychain(credentials, server)
                call.resolve()
            }catch{
                call.reject(error.localizedDescription)
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func deleteCredentials(_ call: CAPPluginCall){
        guard let server = call.getString("server") else {
            call.reject("No server name was provided")
            return
        }
        
        do {
            try deleteCredentialsFromKeychain(server)
            call.resolve()
        }catch {
            call.reject(error.localizedDescription)
        }
    }
    
    
    // Store user Credentials in Keychain
    func storeCredentialsInKeychain(_ credentials: Credentials, _ server: String) throws {
        let query: [String: Any] = [kSecClass as String: kSecClassInternetPassword,
                                    kSecAttrAccount as String: credentials.username,
                                    kSecAttrServer as String: server,
                                    kSecValueData as String: credentials.password.data(using: .utf8)!]
        
        let status = SecItemAdd(query as CFDictionary, nil)
        
        guard status != errSecDuplicateItem else { throw KeychainError.duplicateItem }
        guard status == errSecSuccess else { throw KeychainError.unhandledError(status: status) }
    }
    
    // Update user Credentials in Keychain
    func updateCredentialsInKeychain(_ credentials: Credentials, _ server: String) throws{
        let query: [String: Any] = [kSecClass as String: kSecClassInternetPassword,
                                    kSecAttrServer as String: server]
        
        let account = credentials.username
        let password = credentials.password.data(using: String.Encoding.utf8)!
        let attributes: [String: Any] = [kSecAttrAccount as String: account,
                                         kSecValueData as String: password]
        
        let status = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
        guard status != errSecItemNotFound else { throw KeychainError.noPassword }
        guard status == errSecSuccess else { throw KeychainError.unhandledError(status: status) }
    }
    
    // Get user Credentials from Keychain
    func getCredentialsFromKeychain(_ server: String) throws -> Credentials {
        let query: [String: Any] = [kSecClass as String: kSecClassInternetPassword,
                                    kSecAttrServer as String: server,
                                    kSecMatchLimit as String: kSecMatchLimitOne,
                                    kSecReturnAttributes as String: true,
                                    kSecReturnData as String: true]
        
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status != errSecItemNotFound else { throw KeychainError.noPassword }
        guard status == errSecSuccess else { throw KeychainError.unhandledError(status: status) }
        
        
        
        guard let existingItem = item as? [String: Any],
              let passwordData = existingItem[kSecValueData as String] as? Data,
              let password = String(data: passwordData, encoding: .utf8),
              let username = existingItem[kSecAttrAccount as String] as? String
        else {
            throw KeychainError.unexpectedPasswordData
        }
        
        let credentials = Credentials(username: username, password: password)
        return credentials
    }
    
    // Delete user Credentials from Keychain
    func deleteCredentialsFromKeychain(_ server: String)throws{
        let query: [String: Any] = [kSecClass as String: kSecClassInternetPassword,
                                    kSecAttrServer as String: server]
        
        let status = SecItemDelete(query as CFDictionary)
        guard status == errSecSuccess || status == errSecItemNotFound else { throw KeychainError.unhandledError(status: status) }
    }
}
